### DISCUSSION TITLE
Alternative to multi-threading...

### DISCUSSION THREAD

**Discussion by davidcorbin-atmosera** [libgit2/libgit2#6325] [Q&A] [answered]
Simple multi threaded is not supported/advised as per: 
[(https://github.com/libgit2/libgit2/blob/main/docs/threading.md)]

This leave me in a quandary as to how to do effective (ie fastest possible) deep analysis of repositories. For example I need to examine every file current status across branches and determine when there were merges, of edits made "out of order". The repository is a decent size with about 100K files and an equal number of commits.

Current single threaded takes a few hours to run... :(

**Comment by ethomson** [comment, marked as answer]
If you're not looking at anything in the working directory, if you're just reading commits/trees/blobs/tags (things in the object database), then you should be okay to query the object database from multiple threads.

> The `git_odb` object uses locking internally, and is thread-safe to use from multiple threads simultaneously.

Another option is to have each thread open the repository (each thread having separate `git_repository` objects) and access them independently. I would expect that this might be _slightly_ more performant at the expense of RAM. (Each thread would maintain an independent object cache, which _may_ be more performant depending on your access characteristics.)

**Reply by davidcorbin-atmosera** [reply]
Thanks for the prompt response. I sort of figured those were the basic options, but wanted to mitigate the risk that I was doing something "stupid" or missing a simple opportunity.
